### ISSUE TITLE
Use as client

### ISSUE THREAD

**Issue by eberberich** [Horusiath/yrs-webrtc#7]
In the examples/demo we can see rust code to become aware of changes (of the single doc the server maintains). However we are missing to see an example how to use the code using a rust client. Say we run a warp signaling server (ws, or webrtc) and have clients using WebRrtcProvider from JS y-webrtc, but ALSO a client - written in rust - that connects to a remote document (given a room name) and syncs automatically with the remote documents

A bit like the rust version of the follwing JS

```js
const doc = new Y.Doc()
const provider = new WebrtcProvider('sample', doc, { signaling: ['ws://localhost:8000/signaling'] })
// now changes to doc from events etc .. and propagate them
```

In other words: How a rust client would look like?

**Comment by Horusiath** [comment]
You can see a Rust client (peer) code in the Example section of the readme file: https://github.com/Horusiath/yrs-webrtc/blob/a209523f06aa79c3c5a65c391a9ec2440f00dbc0/README.md?plain=1#L13-L41

**Comment by eberberich** [comment]
We've extended the code - to change a map in the document every few seconds (and showing the map content). And then we start the demo in a different terminal with a different server port - say 800x, but **still connecting the client to port 8000** of the first demo. That is, we can connect multiple clients to the same (first started) signaling server. See code below!

The updates are iterated into the different clients - but not always ... if I add a third or forth client it's not always getting the updates. And: 

Upon connect of a client: We do not get the current version of the (first) document - we only get the document updates with the updates received. Was thinking that the document is shared - and upon connect the client receives the current version from another peer!


```
use log::LevelFilter;
use std::sync::Arc;
use std::time::Duration;
use tokio::time::sleep;
use warp::ws::{WebSocket, Ws};
use warp::{Filter, Rejection, Reply};
use y_sync::awareness::Awareness;
use yrs::updates::decoder::Decode;
use yrs::Update;
use yrs_warp::signaling::{signaling_conn, SignalingService};
use yrs_webrtc::{Error, Room, SignalingConn};

use yrs::Map;
use yrs::MapRef;
use yrs::Transact;
use yrs::Transaction;
use yrs::TransactionMut;
use yrs::types::ToJson;

use tokio::time::interval;
use chrono::Utc;

const STATIC_FILES_DIR: &str = "examples/demo/frontend/dist";

fn show_map(map: MapRef, txn: &Transaction) {
    let state = map.to_json(txn);
    let size = map.len(txn);
    println!("CURRENT MAP STATE A: {:?} (size: {})", state, size);

}

fn show_map_mut(map: MapRef, txn: &TransactionMut) {
    let state = map.to_json(txn);
    let size = map.len(txn);
    println!("CURRENT MAP STATE B: {:?} (size: {})", state, size);
}

async fn run_interval_task(room: Arc<Room>) {
    let mut interval = interval(Duration::from_secs(5));
    loop {
        interval.tick().await;
        println!("add time");
        // Acquire the awareness lock and modify the document
        let awareness = room.awareness().write().await;
        let client_id = awareness.client_id();
        let client_id_string = client_id.to_string();
        let doc = awareness.doc();

        let map = doc.get_or_insert_map("map");
        let mut txn = doc.transact_mut();
        let now = Utc::now();
        let iso_string = now.to_rfc3339();
        map.insert(&mut txn, client_id_string, iso_string);

        show_map_mut(map, &txn);
    }
}

#[tokio::main]
async fn main() -> Result<(), Error> {
    let _ = env_logger::builder()
        .filter_level(LevelFilter::Error)
        .is_test(true)
        .try_init();

    let server = tokio::spawn(signaling_server());
    sleep(Duration::from_secs(1)).await;

    let c1 = Arc::new(SignalingConn::connect("ws://localhost:8000/signaling").await?);
    let r1 = Room::open("sample", Awareness::default(), [c1]);
    let mut pe1 = r1.peer_events().subscribe();

    r1.connect().await?;

    let _ = tokio::spawn(async move {
        while let Ok(e) = pe1.recv().await {
            println!("received peer event: {e:?}");
        }
    });

    let _sub = {
        let a = r1.awareness().write().await;
        a.doc().observe_update_v1(move |_, u| {
            let u = Update::decode_v1(&u.update).unwrap();
            // println!("received update: {u:?}");
        })
    };

    tokio::spawn(run_interval_task(r1));

    server.await?;

    Ok(())
}

async fn signaling_server() {
    let signaling = SignalingService::new();

    let static_files = warp::get().and(warp::fs::dir(STATIC_FILES_DIR));

    let ws = warp::path("signaling")
        .and(warp::ws())
        .and(warp::any().map(move || signaling.clone()))
        .and_then(ws_handler);

    let routes = ws.or(static_files);

    warp::serve(routes).run(([0, 0, 0, 0], 8000)).await; // Each cargo run --example demo runs with a different port number 8000 8001 etc
}

async fn ws_handler(ws: Ws, svc: SignalingService) -> Result<impl Reply, Rejection> {
    Ok(ws.on_upgrade(move |socket| peer(socket, svc)))
}

async fn peer(ws: WebSocket, svc: SignalingService) {
    match signaling_conn(ws, svc).await {
        Ok(_) => println!("signaling connection stopped"),
        Err(e) => eprintln!("signaling connection failed: {}", e),
    }
}
```

**Comment by djmaze** [comment]
I also do not understand how a peer is supposed to become up-to-date on the document state.
